In [1]:
import numpy as np
X=np.array([35, 45, 190, 200, 40, 70, 54, 150, 120, 110])
y=np.array([401, 574, 874, 919, 459, 739, 653, 902, 746, 832])
n=10

In [4]:
b=(np.mean(X*y)-np.mean(X)*np.mean(y))/(np.mean(X**2)-np.mean(X)**2)
b

2.620538882402765

In [5]:
a=(np.mean(y)-b*np.mean(X))
a

444.1773573243596

In [6]:
y_hat=444.1773+2.6205*X
y_hat

array([535.8948, 562.0998, 942.0723, 968.2773, 548.9973, 627.6123,
       585.6843, 837.2523, 758.6373, 732.4323])

In [2]:
import seaborn as sns
X=X.reshape((10,1))
y=y.reshape((10,1))


In [3]:
B=np.dot(np.linalg.inv(np.dot(X.T,X)),X.T @ y)
B


array([[5.88982042]])

In [ ]:
Вроде формулы правельны, но ответ странный. 

Посчитать коэффициент линейной регрессии при заработной плате (zp), используя градиентный спуск (без intercept).

In [28]:
import numpy as np
X=np.array([35, 45, 190, 200, 40, 70, 54, 150, 120, 110])
y=np.array([401, 574, 874, 919, 459, 739, 653, 902, 746, 832])
n=10
B1=0.1


In [29]:
def mse_(B1,y=y,X=X,n=10):
    return np.sum((B1*X-y)**2)/n

In [30]:
alpha = 1e-6
for i in range(2000):
    B1 -=alpha*(2/n)*np.sum((B1*X-y)*X)
    if i%100==0:
        print ('Iteration:{i}, B1={B1}, mse={mse}'.format(i=i,B1=B1, mse=mse_(B1)))

Iteration:0, B1=0.25952808, mse=493237.7212546963
Iteration:100, B1=5.54537842245223, mse=58151.31823171113
Iteration:200, B1=5.868748638669329, mse=56522.97550129376
Iteration:300, B1=5.888531320728348, mse=56516.88130936019
Iteration:400, B1=5.8897415574471985, mse=56516.85850140053
Iteration:500, B1=5.889815595583751, mse=56516.858416040064
Iteration:600, B1=5.889820124983314, mse=56516.85841572062
Iteration:700, B1=5.889820402076462, mse=56516.85841571941
Iteration:800, B1=5.88982041902807, mse=56516.8584157194
Iteration:900, B1=5.889820420065112, mse=56516.85841571941
Iteration:1000, B1=5.8898204201285544, mse=56516.85841571941
Iteration:1100, B1=5.889820420132437, mse=56516.85841571943
Iteration:1200, B1=5.889820420132673, mse=56516.85841571943
Iteration:1300, B1=5.889820420132673, mse=56516.85841571943
Iteration:1400, B1=5.889820420132673, mse=56516.85841571943
Iteration:1500, B1=5.889820420132673, mse=56516.85841571943
Iteration:1600, B1=5.889820420132673, mse=56516.85841571943

Выбираем 600 итерацию с В1=5.88982 

*3. Произвести вычисления как в пункте 2, но с вычислением intercept. Учесть, что изменение коэффициентов должно производиться
на каждом шаге одновременно (то есть изменение одного коэффициента не должно влиять на изменение другого во время одной итерации).

In [49]:
import numpy as np
X=np.array([35, 45, 190, 200, 40, 70, 54, 150, 120, 110])
y=np.array([401, 574, 874, 919, 459, 739, 653, 902, 746, 832])
n=10
B1=0.1
B0=0.1

In [50]:
def mse_(B0,B1,y=y,X=X,n=10):
    return np.sum((B0+B1*X-y)**2)/n

In [51]:
alpha = 1e-5
for i in range(10000000):
    y_pr=B0+B1*X
    B0 -=alpha*(2/n)*np.sum((y_pr-y))
    B1 -=alpha*(2/n)*np.sum((y_pr-y)*X)
    if i%100000==0:
        print ('Iteration:{i}, B0={B0}, B1={B1}, mse={mse}'.format(i=i, B0=B0, B1=B1, mse=mse_(B0,B1)))

Iteration:0, B0=0.1139932, B1=1.6950780000000003, mse=298805.38996099564
Iteration:100000, B0=176.8178756872003, B1=4.588422873669245, mse=24602.65909964911
Iteration:200000, B0=283.1956881772298, B1=3.805435141410038, mse=13044.16516084747
Iteration:300000, B0=347.24756959921285, B1=3.333984975180734, mse=8853.693377127487
Iteration:400000, B0=385.81429152799836, B1=3.0501168504576253, mse=7334.45965488598
Iteration:500000, B0=409.0359675710964, B1=2.8791950478876815, mse=6783.669380682817
Iteration:600000, B0=423.0181317603773, B1=2.77628013758468, mse=6583.983235150707
Iteration:700000, B0=431.43702947702633, B1=2.7143133281816474, mse=6511.588063004361
Iteration:800000, B0=436.50619028818244, B1=2.677002063406801, mse=6485.341570358934
Iteration:900000, B0=439.55841781180436, B1=2.6545363193656377, mse=6475.8260411356305
Iteration:1000000, B0=441.396215664944, B1=2.6410093146294247, mse=6472.376235570253
Iteration:1100000, B0=442.502784858385, B1=2.632864476123355, mse=6471.1255265

KeyboardInterrupt: 

На 4800000 итерации  B0=444.1773573117041, B1=2.620538882495915, mse=6470.414201176654 Их и берем.